### PR TITLE
329 write error condition reporting handling python unit tests

### PR DESCRIPTION
**Pull request by hokiegeek2**
remainder of 329 test cases and client-side error checking enhancements

### CODE REVIEW

File: arkouda/numeric.py [hunk 1]
```diff
@@ -30,7 +30,8 @@ def abs(pda : pdarray) -> pdarray:
         repMsg = generic_msg("efunc {} {}".format("abs", pda.name))
         return create_pdarray(repMsg)
     else:
-        raise TypeError("must be pdarray {}".format(pda))
+        raise TypeError("must be pdarray {}, not {}".\
+                                   format(pda.__class__.__name__))
```
Comment (mhmerrill) [inline arkouda/numeric.py:34]: should there be two arguments to this `format` ?

Comment (hokiegeek2) [reply arkouda/numeric.py:34]: @mhmerrill argh, good call! yes, indeed, will fix

File: arkouda/numeric.py [hunk 2]
```diff
@@ -337,12 +346,16 @@ def histogram(pda : pdarray, bins : int=10) -> pdarray:
...
+    if bins < 1:
+        raise ValueError('bins must be 1 or greater')
     if isinstance(pda, pdarray) and isinstance(bins, int):
         repMsg = generic_msg("histogram {} {}".format(pda.name, bins))
         return create_pdarray(repMsg)
     else:
-        raise TypeError("must be pdarray {} and bins must be an int {}".\
-                        format(pda,bins))
+        raise TypeError("pda must be a pdarray, not a {}".\
+                                  format(pda.__class__.__name__,bins))
```
Comment (mhmerrill) [inline arkouda/numeric.py:358]: two arguments? does `bins` need to be here?

File: arkouda/pdarrayclass.py [hunk 3]
```diff
@@ -1355,6 +1357,10 @@ def std(pda : 'pdarray', ddof : int=0) -> np.float64:
     the estimated variance, so even with ``ddof=1``, it will not be an
     unbiased estimate of the standard deviation per se.
     """
+    if not isinstance(ddof):
```
Comment (mhmerrill) [inline arkouda/pdarrayclass.py:1360]: `ddof` is an instance of what?

File: arkouda/numeric.py (hunk 2 above)
Comment (hokiegeek2) [reply arkouda/numeric.py:358]: jeepers, no, I gotta fix that

File: arkouda/pdarrayclass.py (hunk 3 above)
Comment (hokiegeek2) [reply arkouda/pdarrayclass.py:1360]: int, will fix, man, sorry about that. three errors already

File: arkouda/pdarrayclass.py [hunk 4]
```diff
@@ -1400,15 +1408,18 @@ def mink(pda : 'pdarray', k : int) -> 'pdarray':
...
-        if k == 0:
-            return []
         if pda.size == 0:
-            raise TypeError("must be a non-empty pdarray {} of type int or float".format(pda))
+            raise ValueError("must be a non-empty pdarray {} of type int or float".format(pda))
         repMsg = generic_msg("mink {} {} {}".format(pda.name, k, False))
         return create_pdarray(repMsg)
     else:
-        raise TypeError("must be pdarray {}".format(pda))
+        raise TypeError("must be pdarray, not a {}".format(type(pda)))
```
Comment (mhmerrill) [inline arkouda/pdarrayclass.py:1422]: should this be `format(pda.__class__.__name__)` ?

File: arkouda/pdarrayclass.py [hunk 5]
```diff
@@ -1452,15 +1465,18 @@ def maxk(pda : 'pdarray', k : int) -> 'pdarray':
...
-        if k == 0:
-            return []
         if pda.size == 0:
-            raise TypeError("must be a non-empty pdarray {} of type int or float".format(pda))
+            raise ValueError("must be a non-empty pdarray {} of type int or float".format(pda))
         repMsg = generic_msg("maxk {} {} {}".format(pda.name, k, False))
         return create_pdarray(repMsg)
     else:
-        raise TypeError("must be pdarray {}".format(pda))
+        raise TypeError("must be pdarray, not {}".format(type(pda)))
```
Comment (mhmerrill) [inline arkouda/pdarrayclass.py:1479]: `format(pda.__class__.__name__)` ?

File: arkouda/pdarrayclass.py [hunk 6]
```diff
@@ -1503,15 +1521,18 @@ def argmink(pda : 'pdarray', k : int) -> 'pdarray':
...
-        if k == 0:
-            return []
         if pda.size == 0:
-            raise TypeError("must be a non-empty pdarray {} of type int or float".format(pda))
+            raise ValueError("must be a non-empty pdarray {} of type int or float".format(pda))
         repMsg = generic_msg("mink {} {} {}".format(pda.name, k, True))
         return create_pdarray(repMsg)
     else:
-        raise TypeError("must be pdarray {}".format(pda))
+        raise TypeError("must be pdarray, not {}".format(type(pda)))
```
Comment (mhmerrill) [inline arkouda/pdarrayclass.py:1535]: `format(pda.__class__.__name__)` ?

File: arkouda/pdarrayclass.py [hunk 7]
```diff
@@ -1555,12 +1576,15 @@ def argmaxk(pda : 'pdarray', k : int) -> 'pdarray':
...
-        if k == 0:
-            return []
         if pda.size == 0:
-            raise TypeError("must be a non-empty pdarray {} of type int or float".format(pda))
+            raise ValueError("must be a non-empty pdarray {} of type int or float".format(pda))
         repMsg = generic_msg("maxk {} {} {}".format(pda.name, k, True))
         return create_pdarray(repMsg)
     else:
-        raise TypeError("must be pdarray {}".format(pda))
+        raise TypeError("must be pdarray, not {}".format(type(pda)))
```
Comment (mhmerrill) [inline arkouda/pdarrayclass.py:1590]: `format(pda.__class__.__name__)` ?

File: arkouda/pdarrayclass.py (hunk 5 above)
Comment (hokiegeek2) [reply arkouda/pdarrayclass.py:1479]: yes indeed, will fix. This is an example of why I'd like to work on [540](https://github.com/mhmerrill/arkouda/issues/540)

File: arkouda/pdarrayclass.py (hunk 6 above)
Comment (hokiegeek2) [reply arkouda/pdarrayclass.py:1535]: yes, indeed, will fix

File: arkouda/pdarrayclass.py (hunk 7 above)
Comment (hokiegeek2) [reply arkouda/pdarrayclass.py:1590]: yes, indeed, will fix

File: arkouda/pdarraycreation.py [hunk 8]
```diff
@@ -381,27 +384,45 @@ def linspace(start : int, stop : int, length : int) -> pdarray:
...
+        raise TypeError(('The start parameter must be an int or a scalar that'  +
+                        ' can be parsed to an int, but is a {}'.format(ke)))
     stoptype = resolve_scalar_dtype(stop)
-    stopstr = NUMBER_FORMAT_STRINGS[stoptype].format(stop)
+
+    try: 
+        stopstr = NUMBER_FORMAT_STRINGS[stoptype].format(stop)
+    except KeyError as ke:
+        raise TypeError(('The stop parameter must be an int or a scalar that'  +
+                        ' can be parsed to an int, but is a {}'.format(ke)))
```
Comment (mhmerrill) [inline arkouda/pdarraycreation.py:401]: `format(pda.__class__.__name__)` or  `type()` ?

**Review by mhmerrill — APPROVED** [review]
things look good

**Pull request merged**